x/appchain/subscriber/keeper/keeper.go

```diff
@@ -1,19 +1,139 @@
-func NewKeeper(cdc codec.BinaryCodec, storeKey storetypes.StoreKey) Keeper {
+// NewKeeper creates a new subscriber keeper.
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey storetypes.StoreKey,
+	scopedKeeper commontypes.ScopedKeeper,
+	portKeeper commontypes.PortKeeper,
+	clientKeeper commontypes.ClientKeeper,
+	connectionKeeper commontypes.ConnectionKeeper,
+	channelKeeper commontypes.ChannelKeeper,
+	ibcCoreKeeper commontypes.IBCCoreKeeper,
+) Keeper {
 	return Keeper{
-		cdc:      cdc,
-		storeKey: storeKey,
+		cdc:              cdc,
+		storeKey:         storeKey,
+		scopedKeeper:     scopedKeeper,
+		portKeeper:       portKeeper,
+		clientKeeper:     clientKeeper,
+		connectionKeeper: connectionKeeper,
+		channelKeeper:    channelKeeper,
+		ibcCoreKeeper:    ibcCoreKeeper,
 	}
 }
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// GetPort returns the portID for the IBC app module. Used in ExportGenesis
+func (k Keeper) GetPort(ctx sdk.Context) string {
+	store := ctx.KVStore(k.storeKey)
+	return string(store.Get(types.PortKey()))
+}
+
+// SetPort sets the portID for the IBC app module. Used in InitGenesis
+func (k Keeper) SetPort(ctx sdk.Context, portID string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.PortKey(), []byte(portID))
+}
+
+// IsBound checks if the IBC app module is already bound to the desired port
+func (k Keeper) IsBound(ctx sdk.Context, portID string) bool {
+	_, ok := k.scopedKeeper.GetCapability(ctx, host.PortPath(portID))
+	return ok
+}
+
+// BindPort defines a wrapper function for the port Keeper's function in
+// order to expose it to module's InitGenesis function
+func (k Keeper) BindPort(ctx sdk.Context, portID string) error {
+	capability := k.portKeeper.BindPort(ctx, portID)
+	return k.ClaimCapability(ctx, capability, host.PortPath(portID))
+}
+
+// ClaimCapability allows the IBC app module to claim a capability that core IBC
+// passes to it
+func (k Keeper) ClaimCapability(
+	ctx sdk.Context,
+	cap *capabilitytypes.Capability,
+	name string,
+) error {
+	return k.scopedKeeper.ClaimCapability(ctx, cap, name)
+}
+
+// GetPendingChanges gets the pending validator set changes that will be applied
+// at the end of this block.
+func (k Keeper) GetPendingChanges(
+	ctx sdk.Context,
+) (*commontypes.ValidatorSetChangePacketData, bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.PendingChangesKey())
+	if bz == nil {
+		return nil, false
+	}
+	res := &commontypes.ValidatorSetChangePacketData{}
+	k.cdc.MustUnmarshal(bz, res)
+	return res, true
+}
+
+// SetPendingChanges sets the pending validator set changes that will be applied
+// at the end of this block.
+func (k Keeper) SetPendingChanges(
+	ctx sdk.Context,
+	data *commontypes.ValidatorSetChangePacketData,
+) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(data)
```

Comment:
Handling the marshaling error in any manner other than `panic` is not a good idea. This is because this marshal/unmarshal is extremely key to the entire process and if it fails, we are in deeper waters than you'd think.